File: server/internal/integrations/gitea/repo.go
// Package gitea is the Gitea integration: a bring-your-own Personal/Bot
// Access Token (PAT) connection per workspace, mirrored pull request state,
// and issue auto-link/auto-close from pull_request webhooks. Unlike GitHub
// (an App-marketplace install with ephemeral installation tokens), Gitea has
// no App-install concept, so this mirrors the Slack BYO pattern instead
// (server/internal/integrations/slack/byo_install.go): the workspace admin
// pastes a PAT, we validate it live and store it encrypted at rest.
//
// This package is fully independent of server/internal/handler/github.go and
// the github_* tables — kept that way deliberately so this fork's Gitea
// support never conflicts with upstream GitHub changes.
package gitea

import (
	"net/url"
	"strings"
)

// splitHostAndPath extracts the host and path-with-namespace from a git
// remote URL. Mirrors repocache.splitHostAndPath's logic (kept local rather
// than exported from that package, to avoid coupling this integration to the
// daemon tree) — handles URL form (https://host/path,
// ssh://user@host:port/path) and scp-style shorthand ([user@]host:path).
func splitHostAndPath(rawURL string) (host, path string) {
	if u, err := url.Parse(rawURL); err == nil && u.Scheme != "" && u.Host != "" {
		return u.Host, strings.TrimPrefix(u.Path, "/")
	}
	s := rawURL
	if i := strings.Index(s, "@"); i >= 0 {
		s = s[i+1:]
	}
	if i := strings.Index(s, ":"); i >= 0 {
		return s[:i], s[i+1:]
	}
	return "", s
}

func stripPort(host string) string {
	if i := strings.LastIndex(host, ":"); i >= 0 {
		return host[:i]
	}
	return host
}

// ParseRepoURL extracts (owner, name) from a git remote URL, accepting both
// full URL form and scp-style shorthand. Returns ok=false when the URL
// doesn't carry at least two path segments (owner + repo name).
func ParseRepoURL(rawURL string) (owner, name string, ok bool) {
	_, path := splitHostAndPath(strings.TrimRight(strings.TrimSpace(rawURL), "/"))
	path = strings.TrimSuffix(path, ".git")
	segs := make([]string, 0, 2)
	for _, seg := range strings.Split(path, "/") {
		if seg != "" {
			segs = append(segs, seg)
		}
	}
	if len(segs) < 2 {
		return "", "", false
	}
	return segs[len(segs)-2], segs[len(segs)-1], true
}

// HostMatches reports whether rawURL's host (hostname only, port ignored —
// SSH clone URLs and the HTTPS base URL commonly differ in port) matches the
// configured Gitea instance's hostname. Used to filter workspace.repos down
// to the ones that live on this workspace's connected Gitea instance.
func HostMatches(rawURL, baseURL string) bool {
	host, _ := splitHostAndPath(strings.TrimRight(strings.TrimSpace(rawURL), "/"))
	host = stripPort(host)
	base, err := url.Parse(baseURL)
	if err != nil || base.Hostname() == "" || host == "" {
		return false
	}
	return strings.EqualFold(host, base.Hostname())
}

File: packages/core/gitea/repo.ts
// Mirrors server/internal/integrations/gitea/repo.go's HostMatches — used by
// the UI to decide whether a repo URL belongs to the connected Gitea
// instance (render the Gitea icon / attach as a gitea_repo project resource)
// or should fall back to GitHub. Pure host matching only, no reachability
// check.

function splitHostAndPath(rawUrl: string): { host: string; path: string } {
  try {
    const u = new URL(rawUrl);
    if (u.protocol && u.host) {
      return { host: u.host, path: u.pathname.replace(/^\//, "") };
    }
  } catch {
    // Not a parseable absolute URL (e.g. scp-style shorthand) — fall through
    // to the manual split below.
  }
  let s = rawUrl;
  const at = s.indexOf("@");
  if (at >= 0) s = s.slice(at + 1);
  const colon = s.indexOf(":");
  if (colon >= 0) {
    return { host: s.slice(0, colon), path: s.slice(colon + 1) };
  }
  return { host: "", path: s };
}

function stripPort(host: string): string {
  const idx = host.lastIndexOf(":");
  return idx >= 0 ? host.slice(0, idx) : host;
}

/**
 * Reports whether repoUrl's host (hostname only, port ignored — SSH clone
 * URLs and the HTTPS base URL commonly differ in port) matches the
 * configured Gitea instance's hostname. Accepts the same URL forms as the
 * backend: https://host/path, ssh://user@host:port/path, and scp-style
 * user@host:path shorthand.
 */
export function giteaHostMatches(repoUrl: string, baseUrl: string): boolean {
  const trimmed = repoUrl.trim().replace(/\/+$/, "");
  if (!trimmed || !baseUrl) return false;
  const { host } = splitHostAndPath(trimmed);
  const repoHost = stripPort(host).toLowerCase();
  if (!repoHost) return false;
  try {
    const base = new URL(baseUrl);
    return !!base.hostname && repoHost === base.hostname.toLowerCase();
  } catch {
    return false;
  }
}
